src/meta/ClusterManager.cpp

```diff
@@ -8,40 +8,81 @@
+                               [&](ResultCode innerCode) {
+                                   code = innerCode;
+                                   baton.post();
+                               });
+    baton.wait();
+    return code;
+}
+
+bool ClusterManager::loadOrCreateCluId(NebulaStore* nebulaStore) {
     std::string strClusterId;
-    kvstore->get(kDefaultSpaceId, kDefaultPartId, kClusterIdKey, &strClusterId);
 
-    bool ret = false;
+    ResultCode getCode = nebulaStore->get(kDefaultSpaceId,
+                                          kDefaultPartId,
+                                          kClusterIdKey,
+                                          &strClusterId);
+    if ((getCode != ResultCode::SUCCEEDED) && (getCode != ResultCode::ERR_KEY_NOT_FOUND)) {
+        LOG(ERROR) << "ClusterManager::loadOrCreateCluId get error!";
+        return false;
+    }
+
     if (!strClusterId.empty()) {
         clusterId_ = folly::to<ClusterID>(strClusterId);
+        LOG(INFO) << "ClusterManager::loadOrCreateCluId load from kvstore, clusterId: "
+                  << clusterId_;
         clusterIdIsSet_ = true;
         return true;
     } else {
         createClusterId();
         strClusterId = folly::stringPrintf("%ld", clusterId_);
-        folly::Baton<true, std::atomic> baton;
-        kvstore->asyncMultiPut(kDefaultSpaceId,
-                               kDefaultPartId,
-                               {{kClusterIdKey, strClusterId}},
-                               [&](ResultCode code) {
-                                   ret = code == ResultCode::SUCCEEDED;
-                                   baton.post();
-                               });
-        baton.wait();
-    }
-    return ret;
+        int putTryNum = 3;
+        while (putTryNum > 0) {
+            ResultCode code = kvPut(nebulaStore, strClusterId);
+            --putTryNum;
+            if (code == ResultCode::SUCCEEDED) {
+                LOG(INFO) << "ClusterManager::loadOrCreateCluId kvPut success, clusterId: "
+                          << clusterId_;
+                return true;
+            } else if (code != ResultCode::ERR_LEADER_CHANGED) {
+                LOG(ERROR) << "ClusterManager::loadOrCreateCluId put error!";
+                return false;
+            }
+            auto leaderAddrRet = nebulaStore->partLeader(kDefaultSpaceId, kDefaultPartId);
+            if (!ok(leaderAddrRet)) {
+                LOG(ERROR) << "ClusterManager::loadOrCreateCluId partLeader() error!";
+                return false;
+            }
+            auto leaderAddr = value(leaderAddrRet);
+            bool elecCompleted = (leaderAddr != HostAddr(0, 0));
+            if (elecCompleted) {
+                LOG(INFO) << "ClusterManager::loadOrCreateCluId elecCompleted";
+                return true;
+            } else {
+                LOG(INFO) << "ClusterManager::loadOrCreateCluId not elecCompleted";
+                sleep(5);
```

Comment:
why just this condition need retry?

Reply:
Because ERR_LEADER_CHANGED implies leader election has not completed(leader election need several seconds). Other error code(i.e. ERR_INVALID_ARGUMENT) normally we need not retry(retry is no use).